### ISSUE TITLE
Performance decrease in TensorFlow 1.9 for large graphs

### ISSUE THREAD

**Issue by drasmuss** [tensorflow/tensorflow#20843]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: yes
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 10
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.9.0/1.8.0
- **Python version**: 3.5
- **Bazel version (if compiling from source)**: n/a
- **GCC/Compiler version (if compiling from source)**: n/a
- **CUDA/cuDNN version**: 9.0
- **GPU model and memory**: GTX 980 Ti, 6GB
- **Exact command to reproduce**:

```
import time

import tensorflow as tf

x = tf.zeros((1, 10))

for i in range(5000):
    x = tf.layers.dense(x, units=10)

with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())

    print("a")
    start = time.time()
    sess.run(x)
    print(time.time() - start)

    print("b")
    start = time.time()
    sess.run(x)
    print(time.time() - start)
```

### Describe the problem
On TensorFlow 1.9, this shows:
```
a
17.57940149307251
b
0.09075760841369629
```
On TensorFlow 1.8, this shows:
```
a
11.689541101455688
b
0.08776521682739258
```
So roughly a 50% increase in run time in TensorFlow 1.9.  Note that this only occurs on the first call to `sess.run` ("a" rather than "b"), so I suspect this is some change in the graph optimization step.  So, a couple questions:

1. What changed?  I can't find anything obviously related to this in the release notes.
2. Is this intended/expected (e.g., some new graph optimization steps were added that increase the initial time, but provide benefits in the long run)?  Or is it just a bug?
3. If it is intended, is there some way to optionally disable this new feature if we find that initial build time is dominating the run time?

**Comment by drasmuss** [comment]
I did a git bisect to figure out where the slowdowns occurred, and found 3 commits that contributed to the decreased performance:

https://github.com/tensorflow/tensorflow/commit/2c105ace934edce193669b55b13b64283caa24d7

Runs the graph optimizer pass twice (up from previous value of once), so makes sense that this would take longer.  Can be disabled via `config.graph_options.rewrite_options.meta_optimizer_iterations=RewriterConfig.ONE`.

https://github.com/tensorflow/tensorflow/commit/7e5d24de55572741f10ee7e17247ca195eaadfae

Turns on three new graph optimization steps: dead branch elimination, shape optimization, and remapping.  Shape optimization and remapping can be disabled via `config.graph_options.rewrite_options.shape_optimization/remapping=RewriterConfig.OFF`, but I don't think there is any way to selectively disable dead branch elimination (other than turning off all loop optimizations).

https://github.com/tensorflow/tensorflow/commit/2273b62a769aa477f8d2ef02ca7dee253b8ea7b0

This adds a lot of new logic to the `SymbolicShapeRefiner`, which is what is slowing things down I believe.  I don't think there is any way to disable this.

Note: all three commits are by @benoitsteiner, so he may be able to provide more detail.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by mihi-r** [comment]
I'm experiencing a fairly large slowdown too (on Windows 7 and Python 3.6; CPU) with the release of TensorFlow 1.9 and 1.10. For example, in 1.8 and versions prior, each evaluation step for the [CNN MNIST example model](https://www.tensorflow.org/tutorials/estimators/cnn) takes an average of 13 seconds while in 1.9 and 1.10, it trains for an average of 380 seconds. I hope this is still getting looked over.

**Comment by DavidWiesner** [comment]
I have the same issues after upgrading to tensorflow 1.9, 1.10 or nightly(one our ago) with docker images with Tesla V100 GPU. The global_step/sec slow down from 0.96 to 0.73 with the nasnet-large network trained with a customised version of [tensorflow/models/research/slim](https://github.com/tensorflow/models/tree/master/research/slim)

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @tatatodd: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
@drasmuss  Thanks for interesting investigation. Can you please test this for latest version of TensorFlow and post your findings?. We would like to know if this has been taken care of by latest version.

**Comment by DavidWiesner** [comment]
I had the hope that improvements of `SymbolicShapeRefiner` in commit 9f847cb in the latest nightly fix the issue, but nasnet-large in training is still slow (0.96 -> 0.72 global_step/sec).

**Comment by drasmuss** [comment]
Performance is the same (or slightly worse) on 1.11.0:

```
a
18.11762547492981
b
0.08776497840881348
```

**Comment by smit-hinsu** [comment]
CC @rmlarsen, in case Rasmus has more context about the Grappler changes in https://github.com/tensorflow/tensorflow/issues/20843#issuecomment-406702409

**Comment by tensorflowbutler** [comment]
Nagging Assignees @tatatodd, @sherryy, @azaks2: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by ymodak** [comment]
I see following computation times for various TensorFlow versions and they look pretty descent to me:

TensorFlow Version 1.10 - 
a
14.6700410843
b
0.0304279327393

TensorFlow Version 1.11 - 
a
14.541062355041504
b
0.0331730842590332

TensorFlow Version 1.12.0 -
a
13.787318706512451
b
0.03347063064575195

**Comment by drasmuss** [comment]
The slowdown occurred in version 1.9, so you need to compare 1.9 to 1.8 if you want to see the difference (all the versions you are comparing are post-slowdown).

**Comment by smit-hinsu** [comment]
I talked with @rmlarsen and he mentioned that the regression in https://github.com/tensorflow/tensorflow/commit/2273b62a769aa477f8d2ef02ca7dee253b8ea7b0 has been fixed in later versions but later versions also introduced a few more optimizations.

I don't know details of your use-case but it is possible to disable all graph optimizations with the disable_meta_optimizer option and the time spent on optimizations can be restricted with the meta_optimizer_timeout_ms option.

Another approach could be to use a dummy "warm-up requests" as described in https://github.com/tensorflow/serving/issues/385#issuecomment-290443584.

**Comment by DavidWiesner** [comment]
I created an [example script on gist](https://gist.github.com/DavidWiesner/c6af3238a90b3a8b1209543a9f8d7127#file-flowers-nasnet-sh) with tensorflow models and the large nasnet. This script were run on the official tensorflow docker images.
On my Tesla V100 I got around for tensorflow 1.8-devel-gpu-py3
`INFO:tensorflow:global_step/sec: 1.33334`
With tensorflow 1.9-devel-gpu-py3 a maximum at
`INFO:tensorflow:global_step/sec: 0.89946`

**Comment by smit-hinsu** [comment]
@DavidWiesner That seems like a different issue to me and not related to the graph optimization overhead. If the regression is present in the latest release as well, consider filing a separate issue for that.

**Comment by tensorflowbutler** [comment]
Nagging Assignees @sherryy, @rmlarsen: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by sherryy** [comment]
Closing this issue as longer optimization time is expected with the presence of additional graph optimizations.
